### ISSUE TITLE
Support TZ environment variable

### ISSUE THREAD

**Issue by MiSawa** [Yuri6037/time-tz#2]
Perhaps this is out-of-scope, but I'm wondering if you have a plan to support the [TZ](https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html) (and maybe TZDIR) environment variables so that we can reproduce behavior of, for example,
```bash
$ TZ="FOO+12:34" date "+%Z %z"
FOO -1234
$ TZ=":Cuba" date "+%Z %z"
CST -0500
```

**Comment by Yuri6037** [comment]
Well to start TZ can only be supported by the "system" feature which is what enabled reading timezone from system.

Now to explain the reason behind the current setup: I've got both a Mac, a Linux and a Windows machine (the later as a VM because I'm no longer using windows actively). I've tested this setup on all systems and while I've initially tried TZ I quickly abandoned this idea when I saw that NONE of my Linux machines had this environment variable set by default and that it has to be set by application code which basically falls apart in threaded apps (check time-tz and referenced CVE for more information). Another reason why this is done this way is TZ is non-standard on Mac and Windows. The scope of this project like the scope of all my projects is to be always as cross-platform as possible, which means I needed a solution which would work on most platforms.

It's true that currently the cross-platform part is not entirely finished: on other BSD systems (not Mac/iOS) the current timezone is unreliable because those systems have a tendency to write the current timezone as a file copy instead of a symlink like it's supposed to be, I currently have no long-term solution to this problem.

However that said, if this issue is only about adding parsing of TZ environment variable (not automatic generation) this problem is different: I couldn't find any reliable docs on how this variable works and as I already said none of my linux machines do set it. The only doc I found is from IBM (https://www.ibm.com/docs/en/informix-servers/12.10?topic=products-tz-environment-variable) where it is said that the format is OS defined. That said, granted sufficient documentation I may consider implementing a parsing function.

About TZDIR, I'm not sure what that environment variable does (again no docs). I guess from the name that it probably means some directory where to find zone info. If that is right well there's no need of TZDIR in this project as this project comes with the tzdb built in.

**Comment by MiSawa** [comment]
Thank you for the response. I've researched a bit more on `TZ` environment variables.

### About how standard or non-standard `TZ` is
> I saw that NONE of my Linux machines had this environment variable set by default

Yeah, that'd be expected. This environment variable, as far as I understand, is not to set a system-wide configuration. It exists to override the system-wide configuration for a particular program, like `TZ="JST+9 date"`, so that users can call whatever program with a particular timezone without affecting the timezone used for other programs.

> it has to be set by application code which basically falls apart in threaded apps (check time-tz and referenced CVE for more information).

It usually set by the user (or whatever launch the process like `std::process::Command`) when they start up the app, not via `std::env::set_var`. Unfortunately it can also be set by `std::env::set_var` as you said which caused CVE, but as long as we implement this using `std::env::var` (i.e. not via libc `localtime_r` which internally calls `getenv("TZ")` directly), it should be safe I believe.

> Another reason why this is done this way is TZ is non-standard on Mac and Windows. The scope of this project like the scope of all my projects is to be always as cross-platform as possible, which means I needed a solution which would work on most platforms.

Yeah, it totally make sense to not to implement non-standard things. I'm not an expert in this area, but this `TZ` thing is not the OS-level standards I believe. It's just a thing in `tzset` of POSIX C (and probably other standards?) that most of the programs that handles time-related things, if not all, use. So in fact, apps in OSX also respect what `TZ` says, e.g. https://apple.stackexchange.com/questions/356544 (though can't confirm on my side because I don't have one).
For Windows, I'm more unsure because they're not *nix system how environment variables work there is probably very different. I don't have Windows machine either so can't confirm, though seems like [there are things](https://science.ksc.nasa.gov/software/winvn/userguide/3_1_4.htm) that use `TZ`. And as far as they use a compiler (or actually a C standard library) that support `tzset` function, and the `TZ` lookup were done in the function, the situation should be same.

### About format
[Here](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html#tag_08_03) is a POSIX specification about it, and [here](https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html) is a glibc doc with few more examples. It appeared that the syntax is more complicated, mainly due to the daylight saving time specifiers. And yes, you're right about the new owned struct. That'd be perfect.

Actually apparently what `:Cuba` would do was an implementation-defined (from POSIX's perspective), and [glibc does](https://www.gnu.org/software/libc/manual/html_node/TZ-Variable.html) the directory lookup. `TZDIR` is also a glibc thing according to [a manpage](https://linux.die.net/man/3/timezone). According to what we support (static built-in tzdb), for formats like `:characters` it's probably makes sense to just lookup with the exact match on what we have? (glibc probably supports tz files that users manually generated to the `/usr/share/zoneinfo` or to the `TZDIR`, so technically we can parse the file and create another owned timezone structure to mimic what glibc does, but that's actually probably out-of-scope I guess).

**Comment by Yuri6037** [comment]
Thanks for the docs, I've read the POSIX format and it doesn't seem too hard except for 2 parts:
- Parsing "std" and "dst" as that spec declares the '+' of the offset that follows as optional which means I can't use that as a separator. Currently I have no solution to this problem.
- Parsing start and end dates: creating an algorithm that can parse a Julian day counting leap years (there's nothing like that in time-rs) and an algorithm that can construct a Date from a weekday number (0-6 and not an enum), a week number (1-5) and a month number (1-12), again I didn't find anything usable in Date.

I'll try to open an issue on time-rs around this kind of parsing.

**Comment by MiSawa** [comment]
Thank you for proceeding the issue.
About the first point, I think extracting each parts with regex is possible, but it going to be really messy. I just wrote a parser [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c07cde5feaf6d0ab32214510d87b87f4) (types should be refined and error type should be added), but it's more or less complex. Maybe it's better to use `nom` or some other parser generator.

**Comment by Yuri6037** [comment]
Considering the complexity, I think it might indeed be better to use `nom`. My plan to introduce that feature is as follows:
- Start implementing a parser into a new data structure using `nom`.
- Wait on time-rs response to the opened issue on how to handle the DST start/end.
- Depending on time-rs response possibly roll my own implementation of the algorithms.
- Introduce a new API function `parse_tz` which takes a `&'a str` or a `Into<String>` depending on ability of `nom` to accept working with `&'a str` directly and parse it returning a structure implementing the `Timezone` trait which will connect it to the remaining public API.
- Put all this new code behind the `parse_tz` crate feature; that is to avoid increasing compile time in projects which don't need to handle the `TZ` format.

Note I think for now we can just comply to the full POSIX standard, so excluding GNU extensions.

**Comment by Yuri6037** [comment]
I've started implementation of the tz parser you can check in the `parser-tz` branch.

**Comment by MiSawa** [comment]
Great, looking forward to it :) I'm willing to code-review it if that helps more than the burden.

Also a thing I found while reading specifications: POSIX standard says `std offset [dst [offset][, start [/time] , end [/time]]]` so `std offset dst offset` is a valid specifier(spaces are for clarify) but doesn't mention when to apply dst in this case. The only thing I was able to find is in the very bottom of the [man page of tzset](https://man7.org/linux/man-pages/man3/tzset.3.html) which says

> **/usr/share/zoneinfo/posixrules**
>              When a TZ string includes a dst timezone without anything
>              following it, then this file is used for the start/end
>              rules.  It is in the [tzfile(5)](https://man7.org/linux/man-pages/man5/tzfile.5.html) format.  By default, the
>              zoneinfo Makefile hard links it to the America/New_York
>              tzfile.

I'm not sure what specification specifies it or it's just a GNU extension, but actually the glibc spec says `std offset dst [offset],start[/time],end[/time]` which doesn't seem to allow `std offset dst offset` as oppose to POSIX (while the implementation actually probably allows it). :man_shrugging:

**Comment by Yuri6037** [comment]
> Great, looking forward to it :) I'm willing to code-review it if that helps more than the burden.
> 
> Also a thing I found while reading specifications: POSIX standard says `std offset [dst [offset][, start [/time] , end [/time]]]` so `std offset dst offset` is a valid specifier(spaces are for clarify) but doesn't mention when to apply dst in this case. The only thing I was able to find is in the very bottom of the [man page of tzset](https://man7.org/linux/man-pages/man3/tzset.3.html) which says
> 
> > **/usr/share/zoneinfo/posixrules**
> > When a TZ string includes a dst timezone without anything
> > following it, then this file is used for the start/end
> > rules.  It is in the [tzfile(5)](https://man7.org/linux/man-pages/man5/tzfile.5.html) format.  By default, the
> > zoneinfo Makefile hard links it to the America/New_York
> > tzfile.
> 
> I'm not sure what specification specifies it or it's just a GNU extension, but actually the glibc spec says `std offset dst [offset],start[/time],end[/time]` which doesn't seem to allow `std offset dst offset` as oppose to POSIX (while the implementation actually probably allows it). 🤷‍♂️

Considering it's not specified how this case should be dealt with in the POSIX specification, I think of creating a similar behavior to man tzset: I just access `timezones::db::america::NEW_YORK` and compute new time with it, if there's a `dot_offset > 0` in the returned offset then I apply the custom defined offset as specified in POSIX.

**Comment by Yuri6037** [comment]
I've finally finished a preliminary version with support to parse TZ format according to POSIX standard.

> Great, looking forward to it :) I'm willing to code-review it if that helps more than the burden.

I'd be happy that you code review and also write some tests to ensure my hacks work.

**Comment by Yuri6037** [comment]
This should now be fixed and released in version 0.5.2 on crates.io.
